synapse/storage/databases/main/events.py

```diff
@@ -2152,6 +2152,23 @@ def _store_event_state_mappings_txn(
 
             state_groups[event.event_id] = context.state_group
 
+        # if we have partial state for these events, record the fact. (This happens
+        # here rather than in _store_event_txn because it also needs to happen when
+        # we de-outlier an event.)
```

Comment:
Thanks for the explanation! So the outliers we persist will never be flagged with partial state because we always use `EventContext.for_outlier()` as the context*?

* with the exception of the outlier in `RoomMemberMasterHandler._generate_local_out_of_band_leave` for reasons that are beyond me right now.

Reply:
> * with the exception of the outlier in RoomMemberMasterHandler._generate_local_out_of_band_leave for reasons that are beyond me right now.

Yeah, that looks a bit bogus. It means we end up recording state at the leave event, but also flagging it as an outlier, which seems wrong. Either it shouldn't be an outlier, or we shouldn't record the state at it. I'm going to park this for now though.

Reply:
for the record: I've ended up with a fix to this, at https://github.com/matrix-org/synapse/pull/12154, as it broke in the face of my fixes to https://github.com/matrix-org/synapse/issues/12074.